Conversion to double off by power of 10

Given this JVM snippet:

        val testDouble = 999111.999111
        val testBig = BigDecimal.fromDouble(testDouble)
        val newDouble = testBig.doubleValue(true)
        assertEquals(testDouble, newDouble, 0.0)

assert should work, but assert fails:
java.lang.AssertionError: 
Expected :999111.999111
Actual   :9991119.99111

I haven't tried other scenarios yet, have not finished debugging. This was using 0.2.0. 

The problematic line in BigDecimal.kt is:

                this.significand.longValue(true).toDouble() / double10pow[exponent.toInt()]

In debugger, state of BigDecimal before this line is this:

this = {BigDecimal@859} "9.99111999111E+5"
 precision = 12
 precisionLimit = 0
 roundingMode = {RoundingMode@864} "NONE"
 significand = {BigInteger@865} "999111999111"
 exponent = 5
 scale = -1
 usingScale = false
 decimalMode = null

It looks like this should work but it doesn't.  Still digging.

Environment is AS 4.1, Kotlin 1.4.10

--- a/bignum/src/commonMain/kotlin/com/ionspin/kotlin/bignum/decimal/BigDecimal.kt
+++ b/bignum/src/commonMain/kotlin/com/ionspin/kotlin/bignum/decimal/BigDecimal.kt
@@ -541,7 +541,7 @@ class BigDecimal private constructor(
          */
         fun fromDouble(double: Double, decimalMode: DecimalMode? = null): BigDecimal {
             val doubleString = double.toString()
-            return if (doubleString.contains('.')) {
+            return if (doubleString.contains('.') && !doubleString.contains('E', true)) {
                 parseStringWithMode(doubleString.dropLastWhile { it == '0' }, decimalMode)
             } else {
                 parseStringWithMode(doubleString, decimalMode)
@@ -743,7 +743,7 @@ class BigDecimal private constructor(
             if (floatingPointString.isEmpty()) {
                 return ZERO
             }
-            if (floatingPointString.contains('E') || floatingPointString.contains('e')) {
+            if (floatingPointString.contains('E', true)) {
                 // Sci notation
                 val split = if (floatingPointString.contains('.').not()) {
                     // As is case with JS Double.MIN_VALUE
@@ -1063,7 +1063,7 @@ class BigDecimal private constructor(
     fun divide(other: BigDecimal, decimalMode: DecimalMode? = null): BigDecimal {
         val resolvedDecimalMode = resolveDecimalMode(this.decimalMode, other.decimalMode, decimalMode)
         if (resolvedDecimalMode.isPrecisionUnlimited) {
-            var newExponent = this.exponent - other.exponent
+            val newExponent = this.exponent - other.exponent
             val power = (other.precision * 2 + 6)
             val thisPrepared = this.significand * BigInteger.TEN.pow(power)
             val divRem = thisPrepared divrem other.significand
@@ -1519,14 +1519,12 @@ class BigDecimal private constructor(
          * double multiply or divide to compute the (properly
          * rounded) result.  For large exponent values, use fallback string parse
          */
-        return if (exponent < 0 && exponent.absoluteValue < double10pow.size)
-            toBigInteger().longValue() * double10pow[exponent.absoluteValue.toInt()]
-        else {
-            if (exponent >= 0 && exponent < double10pow.size) {
-                this.significand.longValue(true).toDouble() / double10pow[exponent.toInt()]
+        val l = this.significand.longValue(exactRequired)
+        val divExponent = this.significand.numberOfDecimalDigits().toInt() - 1 - exponent.toInt()
+        return if (l.toDouble().toLong() == l && divExponent >= 0 && divExponent < double10pow.size) {
+                l.toDouble() / double10pow[divExponent]
             } else
                 this.toString().toDouble()
-        }
     }
 
     /**

--- a/bignum/src/commonTest/kotlin/com/ionspin/kotlin/bignum/decimal/BigDecimalNarrowingTest.kt
+++ b/bignum/src/commonTest/kotlin/com/ionspin/kotlin/bignum/decimal/BigDecimalNarrowingTest.kt
@@ -28,6 +28,7 @@ import kotlin.test.assertFailsWith
  */
 
 class BigDecimalNarrowingTest {
+    private val testDouble = 0.000000000000999111999111
 
     @Test
     fun doubleValueTest() {
@@ -39,6 +40,17 @@ class BigDecimalNarrowingTest {
         assertEquals(Double.MIN_VALUE, dub.doubleValue(true))
         assertFailsWith<ArithmeticException> {
             dub.divide(BigDecimal.TEN).doubleValue(true)
+        }
+
+        /*
+         * Test a range of exponents on the testDouble value.  
+         */
+        var dubIn = testDouble
+        for (factor in 0..22) {
+            dubIn *= 10.0
+            val bd = BigDecimal.fromDouble(dubIn)
+            val dubOut = bd.doubleValue(true)
+            assertEquals(dubIn, dubOut)
         }
     }
 

